### DISCUSSION TITLE
Avif

### DISCUSSION THREAD

**Discussion by rosell-dk** [rosell-dk/webp-express#605] [Ideas]
Lets talk about AVIF and how best to get there!

Here are some options I'm considering:
1. Simply make a new plugin "Avif Express", which is essentially this plugin, but modified to generate and serve AVIF instead. Actually, such modification has already been done (but not published). See here: https://wordpress.org/support/topic/avif-support-is-here/

2. Modify the plugin so it both handles Avif and Webp. However, this is a great and time consuming task and I have limited time at disposal and other cool projects in queue

3. Create several new plugins, which are more specialized in their purpose. This way, only getting half way will still have value. 

I'm leaning towards 3. Or maybe first 1 and then 3

Here is a division I'm thinking of:

A:  A plugin for creating image redirects in .htaccess. I'm visualising a builder. You add a redirection and configure it, and will be able to add multiple redirections. A redirection could for example be: "Redirect PNG, JPEG and WEBP images in uploads folder to .avif, providing that there is an avif with same name in the same folder (".avif" prepended to filename)" or "Redirect WEBP images in the plugins folder to AVIF images found in a given folder (".avif" replaces file extension)". I will be able to use much of the code in WebP Express for this. WebP Express has a quite robust implementation of generating .htaccess rules, which takes into consideration what .htaccess features the server supports. This way, the plugin will be useful argumenting any image conversion plugin.

B: A plugin for converting images into multiple formats, including webp and avif. Focus: Purely conversion. I already have created a library, [image-convert](https://github.com/rosell-dk/image-convert), which supports converting from multiple formats into multiple formats. So it is just the UI that needs to be written (and maybe/maybe not also triggers, such as when an image is uploaded). A version without UI would be very quick to implement, but might have its use. The idea here is just to provide a function for converting, which other plugins can use. Later, WP-CLI interface can be quickly added. And later yet, a UI can be provided, maybe as another plugin!

C: A plugin for rewriting HTML to add avif / webp images as alternatives in "picture" elements.

I'm in doubt whether to start with 3A or 3B and which of these would be most useful...

**Comment by spikerrbg** [comment]
when try to add avif file write "this file not support to upload"

**Reply by rosell-dk** [reply]
WebP Express does not support Avif. This discussion is about the road to achieve it - either in WebP Express or in a new plugin.

In some way, the name "WebP Express" kind of suggests that the plugin is focused on webp and one would perhaps not expect it to do AVIF too. This would be an argument not to implement it in WebP Express, but rather make a new plugin, ie "Avif Express". But perhaps it would be a better idea to split up the functionality into several smaller plugins, which can be combined.

If you are looking for Avif Support, there is actually a plugin called "Avif support", which adds support for Avif in Wordpress (upload and thumbnail generation support)

**Reply by audioscavenger** [reply]
You need to enable AVIF support first. 2 plugins offer this: `AVIF Support [GrandPlugins]` and `AVIF Support | AVIF Uploader`
Both use similar code explained [here](https://grandplugins.com/how-to-upload-avif-images-in-wordpress/) - ie you can enable AVIF with just a php snippet (given that your server has phpImagick indeed)

**Comment by audioscavenger** [comment]
I vote for (1), hands down. Easiest, fastest, job done. Another plugin has the same name `AVIF express` unfortunately.
Conversion with image-convert is another big project, that seems nicer actually:
* Once I paste images as png (I use `The Paste`, another big need for WP) or upload jpegs or whatnot, I get all the thumbnails paired with an avif immediately, including the full size image. Therefore I don't need the originals anymore.
* Transparent png conversion is current;y buggy, the transparent layer is lost. No idea why.
* Conversion takes a toll on my server each time I paste a picture, i wait like 5 seconds. Because WP builds the thumbnails in the original format, and the avifs at the same time.
Would this conversion project of yours process files during paste/upload as well?
In a dream world, I would like your AVIF express to convert only the full size uploaded, and build the thumbnails from there. I don't care about the originals on the server, i have them on my desktop. So does everyone else...?

(2) is absolutely not needed imo. Despite my limited experience with AVIF, the literature and the quality results I see, all point to a future where AVIF replaces JPEG, not Webp. 
* Webp is closed source from google and the quality at equal size to avif is atrocious. I found some blogs comparing quality, and comparing results of a 100 times re-compression in each format: colors are totally lost with webp.
* Webp had its chance but missed. All browsers and mobile versions support Avif, FF since v92, etc. It's been a while. 
* Avif is open source and can be animated from what I understand, can replace gif as well. 
* Once you go avif, never go back.
* WP supports WebP natively, although it does not convert.
I just gave up on Webp completely, for all these reasons.

(3) not sure what you are thinking about... more plugins? to do what exactly? 

The conversion plugin can be useful when you have tons of pictures, yes. Since every picture is doubled with its avif pair. So in the end, you have a choice to make, since there is only 24h in a day!
Conversion plugin and Avif express have two different purposes I think. I don't mind keeping the originals next to their avif pairs, because one day WP may switch to nginx and all your htaccess rules are lost. Or maybe I'm just naive, and avif should be vanilla because of CDNs? I heard a rumor that they don't like when the mime changes.

in the end, i'd love to help but not sure what I can do. you webp express is such a large project it gave me headaches just parsing the folders.

**Reply by audioscavenger** [reply]
hey how are you doing? 
I was swamped on other projects that I just finished. I also am starting working on a house I am buying, but in between I should have time to help assisting on that AVIF project.

So far, it's been more then a month that i work with avif and webp.
Webp is still needed for ComfyUI unfortunately, and still not mainstream. ComfyUI dev team have not aknowledge the PR I submited to get it working, yet.
But one thing is clear: the more I use avif, the more I love it. The color rendering is as bad as webp, but the artifacts are so much less visible and overall quality is so much higher.
I have setteled at a default of 60% quality for avif and 75% for webp. This not only gives better results then webp, but also 50% smaller files overall.

The only big big issue is indeed the processing power needed to generate avif. I started converting my png workflow exports to avif and let's say, i'm not too stoked by the time it takes. Large images of 5000x3000 take almost 30s on an i9 9700F ... that's not good at all.
Not sure how but whichever direction we take on the project, we have to make it clear to end users that they need powerfull web servers or they will face constant conversion errors and aborts. Webp is much more forgiving and encoding is so much faster, maybe even faster than png.
Also, users will need to increase the php `max_execution_time` beyond 30s, to handle large images. This is not possible on wordpress.com hosted sites and similar. This already limits the scope of the audience.

Personally, I don't want webp to become the next jpeg. it's currently equaly as niche as webp, and i will do everything I can to push avif versus webp.

thoughts? comments?

**Comment by nisbet-hubbard** [comment]
The issue with option 1 is that AVIF isn’t a superior format across the board. WebP performs better when the source image is PNG and we want to keep it lossless: https://github.com/AOMediaCodec/av1-avif/issues/111. So, ideally, we hope there’s an option in the new plugin to default to AVIF for JPEG, and to WebP for PNG (may also be quicker to implement?).

In our use case, because we just bulk convert once and cache the altered HTMLs periodically, 3B+C are the useful parts. With the difference that what we really count on in C is the ability to replace URLs according to whether an image is converted to AVIF or WebP during B.

Considering the limited time @rosell-dk can devote to this, perhaps a first implementation of C could simply be to alter URLs to either `.jpe?g.avif` or `.png.webp`, which tends to be the most performant combination, and add other options later as enhancements.

**Comment by medoeldin** [comment]
Hello Rosell, my developer installed the WebP plugin and it appears they did something which killed the images on my website potteryworld.com. Is this correctable or are the images permanently lost?  Thank you!

**Comment by bobhuf** [comment]
Hello, it's been a while, I was just wondering if avif support has been terminated. Some other plugins have jumped into the space, even a plugin by WordPress Performance Team called [Modern Image Formats](https://wordpress.org/plugins/webp-uploads/ exists now but none have the plug-and-play experience webp-express offers. An avif-express plugin would still be a great addition to the WP ecosystem.
